WIP: Add ant ivy scripts

Tobias321 opened the pull request:
resolves #45, resolves #61

cagix commented on build.xml:
```diff
@@ -0,0 +1,49 @@
+        <delete dir="lib" />
+    </target>
+
+    <target name="compile">
```
shouldn't this depend on "`resolve`"?

cagix commented on build.xml:
```diff
@@ -0,0 +1,49 @@
+
+    <target name="clean">
+        <delete dir="${build.dir}" />
+        <delete dir="lib" />
```
could you use a property for `lib` as well?

cagix commented on build.xml:
```diff
@@ -0,0 +1,49 @@
+
+    <target name="compile">
+        <mkdir dir="${classes.dir}" />
+        <javac srcdir="${src.dir}" destdir="${classes.dir}" includes="lib/*.jar" />
```
not quite sure if this works. `includes` is usually used to define `filesets`, but here you'd need some `classpath` or something?

cagix commented on build.xml:
```diff
@@ -0,0 +1,49 @@
+
+    <target name="clean-build" depends="clean,jar" />
+
+    <target name="main" depends="clean,jar" />
```
??? why not declare `default=clean-build`?  or leave this and remove `clean-build`?

cagix commented on build.xml:
```diff
@@ -0,0 +1,49 @@
+    <available classname="org.apache.ivy.Main" property="ivy.installed" />
+
+    <target name="install-ivy" description="Install ivy" unless="ivy.installed">
+        <mkdir dir="${user.home}/.ant/lib" />
```
~~wouldn't it be better to download all dependencies to `"${lib}""`?~~

edit: this is about installing ivy and not downloading dependencies. sry, misread. actually, this is really neat, if this works ...

cagix commented on build.xml:
```diff
@@ -0,0 +1,49 @@
+
+    <!-- === target: resolve === -->
+    <target name="resolve" description="--> retrieve dependencies with Ivy">
+        <ivy:retrieve pattern="${build.dir}/lib/[artifact]-[revision](-[classifier]).[ext]" />
```
i'd prefer the lib-folder outside of the build-folder.

cagix commented on ivy.xml:
```diff
@@ -0,0 +1,17 @@
+        <dependency org="com.badlogicgames.gdx" name="gdx-platform" rev="1.10.0"/>
+        <dependency org="com.badlogicgames.gdx" name="gdx-box2d" rev="1.10.0"/>
+        <dependency org="com.badlogicgames.gdx" name="gdx-box2d" rev="1.10.0"/>
+         <dependency org="com.badlogicgames.gdx" name="gdx-freetype" rev="1.10.0"/>
```
please re-check indentation ;)

cagix commented on ivy.xml:
```diff
@@ -0,0 +1,17 @@
+        <dependencies>
+        <dependency org="com.google.code.gson" name="gson" rev="2.7"/>
+        <dependency org="com.badlogicgames.gdx" name="gdx-ai" rev="1.8.1"/>
+        <dependency org="com.badlogicgames.gdx" name="gdx-backend-lwjgl3" rev="1.10.0" e:classifier="natives-desktop" />
```
not sure about `e:classifier`? according to https://mvnrepository.com/artifact/com.badlogicgames.gdx/gdx-backend-lwjgl3/1.10.0 this entry should be `<dependency org="com.badlogicgames.gdx" name="gdx-backend-lwjgl3" rev="1.10.0"/>`?

cagix reviewed (changes requested):
this looks quite nice :)

see some comments ... are you sure you want to commit the `ivy-2.5.0.jar`?  ...

Tobias321 commented:
@cagix He builds the (fat) jar file with `jar`. But when i include the resulting `pmdungeon.jar` and try to run the game i get an exception:

```
Exception in thread "main" java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
	at de.fhbielefeld.pmdungeon.desktop.DesktopLauncher.run(Unknown Source)
	at Main.main(Main.java:51)
Caused by: java.lang.ClassNotFoundException: org.lwjgl.glfw.GLFWErrorCallbackI
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:606)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:168)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 2 more
```

this could be, because `e:classifier="natives-desktop"` has no effect, i forget some dependencies, or something is wrong with `<zipgroupfileset dir="${lib.dir}" includes="*.jar" />`.

cagix commented:
@Tobias321 hmmm, not quite sure that this is connected with naming.  could your please take your jar and the one @AMatutat provides ("releases"), unzip both and compare the contents of the folders?

Tobias321 commented:
@cagix Something is here different:

![image](https://user-images.githubusercontent.com/85501570/122011081-0668a880-cdbc-11eb-9e79-2992be535bb0.png)

(on the left Andre, on the right me)

Andre:
```
JarIndex-Version: 1.0

lwjgl-natives-macos.jar
META-INF
META-INF/versions
META-INF/versions/9
macos
macos/x64
macos/x64/org
macos/x64/org/lwjgl
```

my:
```
JarIndex-Version: 1.0

lwjgl-opengl-natives-windows.jar
META-INF
META-INF/versions
META-INF/versions/9
windows
windows/x64
windows/x64/org
windows/x64/org/lwjgl
windows/x64/org/lwjgl/opengl
```

`opengl`?

Tobias321 commented:
i found out, what's going wrong...

`zipgroupfileset` opens and includes all jar files, that are in `dir`. but it not includes the dll files, etc., that are also needed. unfortunately i think, there is no way to do this with ant. i tried it with `zipgroupfileset`, `zipfileset` and `fileset`.

only thing is to include a jar inside a jar as it is, but then java cannot find the content of the inner jar, even when the `Class-Path` is set in the manifest: https://stackoverflow.com/questions/12357136/reference-jars-inside-a-jar

so, in a jar file, only classes, dlls, etc. ("binaries") can be processed - not other jars...

cagix commented:
> so, in a jar file, only classes, dlls, etc. ("binaries") can be processed - not other jars...

so we need to unwrap those jar's and re-wrap the content plus our classes etc. into a new jar. i think, this is just what @AMatutat and the IntelliJ of his do when creating a release ...

Tobias321 commented:
> hmmm, perhaps i got the meaning of this "fetch&update" button wrong after all ... it seems, you just did merge `master` into your feature branch?

that was GitHub Desktop 😢 maybe i should have used strict the "fetch&update" button

Tobias321 commented:
> so we need to unwrap those jar's and re-wrap the content plus our classes etc. into a new jar.

some contents are missing in the jar. i don't know why. i think, ant recognizes only class files in dependency jars.

cagix commented:
> > hmmm, perhaps i got the meaning of this "fetch&update" button wrong after all ... it seems, you just did merge `master` into your feature branch?
> 
> that was GitHub Desktop 😢 maybe i should have used strict the "fetch&update" button

hmmm, i'm trying to understand all those buttons presented to me in the web gui ... usually i just use the cmd line and the good ol' "`git fetch upstream && git rebase upstream/master feature && git push origin feature`" procedure 😀

Tobias321 commented:
i hope, the git history is still "intact" now... next time i use either only the web gui or cmd ...

AMatutat commented:
@Tobias321 still one open change request. Pls fix or mark as done.

Tobias321 commented:
i had added repos maven and central as resolvers in `ivysettings.xml`.

i had also added an `ant compile-run` target, to compile the sources and to run the application.

Anyway, i still get this error:

```
compile-run:
     [java] java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
     [java]     at org.apache.tools.ant.taskdefs.ExecuteJava.execute(ExecuteJava.java:194)
...
```

Can anyone reproduce this error? I don't know what to do next.

Tobias321 commented:
> @Tobias321 still one open change request. Pls fix or mark as done.

how can i do that?

Tobias321 commented:
> are you sure you want to commit the `ivy-2.5.0.jar`? ...

@cagix no, but otherwise, you have to download Ivy manually to run Ant...

AMatutat commented:
> how can i do that?

@Tobias321  Normally you should have a "resolve conversation" button for that. Guess because the review is outdated, you have to dismiss the review. 

![image](https://user-images.githubusercontent.com/32962412/122461752-0fbc6580-cfb4-11eb-90e1-eea773052e77.png)

cagix commented:
> > are you sure you want to commit the `ivy-2.5.0.jar`? ...
> 
> @cagix no, but otherwise, you have to download Ivy manually to run Ant...

but didn't you have a target that "installed" apache ivy?

Tobias321 commented:
> > > are you sure you want to commit the `ivy-2.5.0.jar`? ...
> > 
> > 
> > @cagix no, but otherwise, you have to download Ivy manually to run Ant...
> 
> but didn't you have a target that "installed" apache ivy?

AFAIK, this only checks, if apache ivy is already installed, and abort compilation if not... afaik, it not installs ivy... :(

Tobias321 commented:
> > > > are you sure you want to commit the `ivy-2.5.0.jar`? ...
> > > 
> > > 
> > > @cagix no, but otherwise, you have to download Ivy manually to run Ant...
> > 
> > 
> > but didn't you have a target that "installed" apache ivy?
> 
> AFAIK, this only checks, if apache ivy is already installed, and abort compilation if not... afaik, it not installs ivy... :(

i found out, how to install it: `https://ant.apache.org/ivy/history/2.4.0/install.html`

give me a few minutes

Tobias321 commented:
@AMatutat thanks for renaming

please run `ant init-ivy` and then run `ant compile-run`, at this point i ever get the error mentioned above.

AMatutat commented:
Thats my error. 

```
[ivy:retrieve] :: problems summary ::
[ivy:retrieve] :::: WARNINGS
[ivy:retrieve]          module not found: com.badlogicgames.gdx#gdx;1.10.0
[ivy:retrieve]  ==== my-maven: tried
[ivy:retrieve]    https://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx/1.10.0/gdx-1.10.0-natives-desktop.pom
[ivy:retrieve]    -- artifact com.badlogicgames.gdx#gdx;1.10.0!gdx.jar:
[ivy:retrieve]    https://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx/1.10.0/gdx-1.10.0-natives-desktop.jar
[ivy:retrieve]  ==== central: tried
[ivy:retrieve]    http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx/1.10.0/gdx-1.10.0-natives-desktop.pom
[ivy:retrieve]    -- artifact com.badlogicgames.gdx#gdx;1.10.0!gdx.jar:
[ivy:retrieve]    http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx/1.10.0/gdx-1.10.0-natives-desktop.jar
[ivy:retrieve]          ::::::::::::::::::::::::::::::::::::::::::::::
[ivy:retrieve]          ::          UNRESOLVED DEPENDENCIES         ::
[ivy:retrieve]          ::::::::::::::::::::::::::::::::::::::::::::::
[ivy:retrieve]          :: com.badlogicgames.gdx#gdx;1.10.0: not found
[ivy:retrieve]          ::::::::::::::::::::::::::::::::::::::::::::::
[ivy:retrieve]
[ivy:retrieve] :::: ERRORS
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/org/sonatype/oss/oss-parent/7/oss-parent-7.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/google/code/gson/gson-parent/2.8.6/gson-parent-2.8.6.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx/1.10.0/gdx-1.10.0-natives-desktop.pom
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx/1.10.0/gdx-1.10.0-natives-desktop.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/org/sonatype/oss/oss-parent/5/oss-parent-5.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-parent/1.10.0/gdx-parent-1.10.0.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/apple/AppleJavaExtensions/1.4/AppleJavaExtensions-1.4-sources.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/apple/AppleJavaExtensions/1.4/AppleJavaExtensions-1.4-javadoc.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-platform/1.10.0/gdx-platform-1.10.0-sources.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-platform/1.10.0/gdx-platform-1.10.0-javadoc.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-box2d-parent/1.10.0/gdx-box2d-parent-1.10.0.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-box2d-platform/1.10.0/gdx-box2d-platform-1.10.0-sources.jar
[ivy:retrieve]  SERVER ERROR: HTTPS Required url=http://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-box2d-platform/1.10.0/gdx-box2d-platform-1.10.0-javadoc.jar
[ivy:retrieve]
[ivy:retrieve] :: USE VERBOSE OR DEBUG MESSAGE LEVEL FOR MORE DETAILS
```

Tobias321 commented:
> Looks like a wrong dependencie to me.

he wants to have https, i think. that's why i have added `root="https://repo1.maven.org/maven2/"` to `ivysettings.xml`...

Tobias321 commented:
can you re-try?

AMatutat commented:
Okay, now i also got
```
compile-run:
     [java] java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
     [java]     at org.apache.tools.ant.taskdefs.ExecuteJava.execute(ExecuteJava.java:198)
     [java]     at org.apache.tools.ant.taskdefs.Java.run(Java.java:832)
     [java]     at org.apache.tools.ant.taskdefs.Java.executeJava(Java.java:226)
     [java]     at org.apache.tools.ant.taskdefs.Java.executeJava(Java.java:135)
     [java]     at org.apache.tools.ant.taskdefs.Java.execute(Java.java:108)
     [java]     at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:293)
     [java]     at jdk.internal.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
     [java]     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     [java]     at java.base/java.lang.reflect.Method.invoke(Method.java:564)
     [java]     at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
     [java]     at org.apache.tools.ant.Task.perform(Task.java:352)
     [java]     at org.apache.tools.ant.Target.execute(Target.java:437)
     [java]     at org.apache.tools.ant.Target.performTasks(Target.java:458)
     [java]     at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1406)
     [java]     at org.apache.tools.ant.Project.executeTarget(Project.java:1377)
     [java]     at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
     [java]     at org.apache.tools.ant.Project.executeTargets(Project.java:1261)
     [java]     at org.apache.tools.ant.Main.runBuild(Main.java:857)
     [java]     at org.apache.tools.ant.Main.startAnt(Main.java:236)
     [java]     at org.apache.tools.ant.launch.Launcher.run(Launcher.java:287)
     [java]     at org.apache.tools.ant.launch.Launcher.main(Launcher.java:112)
     [java] Caused by: java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
     [java]     at de.pmdungeon.desktop.DesktopLauncher.run(Unknown Source)
     [java]     at de.pmdungeon.desktop.Main.main(Unknown Source)
     [java]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     [java]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
     [java]     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     [java]     at java.base/java.lang.reflect.Method.invoke(Method.java:564)
     [java]     at org.apache.tools.ant.taskdefs.ExecuteJava.run(ExecuteJava.java:221)
     [java]     at org.apache.tools.ant.taskdefs.ExecuteJava.execute(ExecuteJava.java:154)
     [java]     ... 20 more
     [java] Caused by: java.lang.ClassNotFoundException: org.lwjgl.glfw.GLFWErrorCallbackI
     [java]     at org.apache.tools.ant.AntClassLoader.findClassInComponents(AntClassLoader.java:1386)
     [java]     at org.apache.tools.ant.AntClassLoader.findClass(AntClassLoader.java:1335)
     [java]     at org.apache.tools.ant.AntClassLoader.loadClass(AntClassLoader.java:1096)
     [java]     at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
     [java]     ... 28 more
     [java] Java Result: -1
```

Tobias321 commented:
Ok, seems that Ant is retrieving natives for macos:

![image](https://user-images.githubusercontent.com/85501570/122533833-403ae880-d022-11eb-9926-77f32b572347.png)

Anyway, i can't solve this at the moment.

cagix commented on build.xml:
```diff
@@ -0,0 +1,80 @@
+<project xmlns:ivy="antlib:org.apache.ivy.ant" name="pmdungeon" basedir="." default="clean-build">
+    <property name="src.dir.1" value="core/src" />
```
do we really have `core` and `desktop` anymore?

cagix replied:
hmmm, ok. #65 is not merged yet ...

cagix commented on build.xml:
```diff
@@ -0,0 +1,80 @@
+    <property name="src.dir.1" value="core/src" />
+    <property name="src.dir.2" value="desktop/src" />
+    <property name="build.dir" value="build" />
+    <property name="lib.dir" value="mylibs" />
```
"mylibs"? isn't "libs" more common?

cagix commented on build.xml:
```diff
@@ -0,0 +1,80 @@
+        </jar>
+    </target>
+
+    <target name="run" depends="jar" description="target to resolve all dependencies, to compile all sources and to build and run the fat jar ('main-class' is used as entry point)">
```
one target to rule them all ;)

do we really need `compile-run` and `run`? and if so, wouldn't be `compile-run` the default "run" target, i.e. shouldn't it get the shorter name `run`?

Tobias321 replied:
i could rename `run` to `jar-run`?

cagix replied:
@AMatutat what's your opinion on this?

AMatutat replied:
@cagix sounds good

cagix commented on build.xml:
```diff
@@ -0,0 +1,80 @@
+
+    <target name="clean-build" depends="clean,jar" description="default target to clean the build directory, to resolve all dependencies, to compile all sources and to build the fat jar (cleans not the mylibs directory)" />
+
+    <property name="ivy.install.version" value="2.1.0-rc2" />
```
i'd prefer a single spot to define all properties, not to scatter this across the script ...

Tobias321 replied:
ok 👍

cagix commented on ivy.xml:
```diff
@@ -0,0 +1,14 @@
+<ivy-module xmlns:e="http://ant.apache.org/ivy/extra" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" version="2.0" xsi:noNamespaceSchemaLocation="http://ant.apache.org/ivy/schemas/ivy.xsd">
+    <info organisation="pmdungeon" module="pmdungeon" />
+    <dependencies>
+        <dependency org="com.google.code.gson" name="gson" rev="2.8.6" />
```
@AMatutat for didactic reasons, i think it's good that we provide all three build scripts (ant, maven, gradle). but i'm afraid we really have to be careful with the version numbers, so that they are always in sync.

do you see a chance to somehow define and maintain the version numbers in a common file?

AMatutat replied:
#69

Tobias321 commented on build.xml:
```diff
@@ -0,0 +1,80 @@
+    </target>
+
+    <target name="compile-run" depends="compile">
+        <java classpath="${classes.dir}" classpathref="classpath" classname="${main-class}">
```
yes, one declaration is better, please see b2e7c4f

Tobias321 commented on desktop/src/de/pmdungeon/desktop/Main.java:
```diff
@@ -0,0 +1,38 @@
+
+import de.pmdungeon.game.Controller.MainController;
+
+public class Main extends MainController {
```
is it bad, to add a main class (at this point), for easier build testing? ... :(

cagix replied:
define "bad" :)

i think of the use case: students should pull this repo and use it as a starting point, and implementing this class was one of the first tasks ... so, in a way, this would be a kind of anticipation of the solution ...

on the other hand, it is customary to introduce only one thing per pr ...

cagix replied:
@Tobias321 I understand that you need this class to test the build scripts, but maybe you can just keep the class local without making it part of the PR?

Tobias321 replied:
ok :)

as you described, one issue = one pr + noting the use case (or don't steal the students their tasks :) )

Tobias321 commented on README.md:
```diff
@@ -17,6 +17,13 @@ Dieses Dokument liefert einen Einstieg in das PM-Dungeon. Es erläutert die Inst
 - Integrieren Sie die `pmdungeon.jar` als Library in Ihrem Projekt
 - Kopieren Sie den `assets`-Ordner in Ihr Projektverzeichnis
 
+### Installation mit Apache Ant
```
i have to rewrite this part... it is also a bit "talkative/redundant"

cagix commented on README.md:
```diff
@@ -17,6 +17,13 @@ Dieses Dokument liefert einen Einstieg in das PM-Dungeon. Es erläutert die Inst
 
+### Installation mit Apache Ant
+
+- Um eine `jar`-Datei zu erstellen, die Sie in Ihr Projekt integrieren können, rufen Sie das ant target `ant jar` auf. Ant wird dann die Abhängigkeiten herunterladen und die `pmdungeon.jar`-Datei in dem Verzeichnis `build/jar` erstellen.
```
btw: that means, please replace "ant target" w/ "Ant-Target" ...

Tobias321 commented:
should i try rebasing this branch to one commit and push it? i'm not sure what then happens with this pr. and with the git history.

if there is an accident, i still can use another branch...

AMatutat commented:
> should i try rebasing this branch to one commit and push it?

Why should you?

Tobias321 commented:
> > should i try rebasing this branch to one commit and push it?
> 
> Why should you?

ok, then not 😉 because of add and delete Main... but, all ok 🙂

cagix reviewed (approved):
lgtm 👍 

however, i've only done a code review, but not tested the script live ...

cagix commented:
> > > should i try rebasing this branch to one commit and push it?
> > 
> > 
> > Why should you?
> 
> ok, then not 😉 because of add and delete Main... but, all ok 🙂

if the branch is being merged, in the end only the difference between the start and the end will matter ... 

rebasing is of interest if you have merge conflicts or want to have the last changes from the target branch in your branch or want to have a linear history after the merge ...

Tobias321 commented:
but there is still the problem with
`java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI`
when running this with ant...

i've searched for that on the internet, and nobody has the same error :(

Tobias321 commented:
> if the branch is being merged, in the end only the difference between the start and the end will matter ...

so git is "flattening the curve" when merging a feature branch, that's nice :)

AMatutat commented:
> but there is still the problem with
java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
when running this with ant...

I guess we are missing a dependency here, the question is which one.

Tobias321 commented:
> > but there is still the problem with
> > java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
> > when running this with ant...
> 
> I guess we are missing a dependency here, the question is which one.

Is it conceivable, that different versions of LibGDX are in use? Maybe that's the point.

AMatutat commented:
I merged #65, so check if this need some adjustments.

Tobias321 commented:
> I merged #65, so check if this need some adjustments.

Thanks. There is still the same problem with `GLFWErrorCallbackI` on windows. 😢

Tobias321 commented:
`[ivy:retrieve]  com.badlogicgames.gdx#gdx;1.9.8 (needed by [com.badlogicgames.gdx#gdx-ai;1.8.2]) conflicts with com.badlogicgames.gdx#gdx;1.10.0 (needed by [pmdungeon#pmdungeon;working@...])` that is the conflict we have...

cagix commented:
hmmm, i still don't understand, why we have to define each dependency individually... all the specialised parts depend on some path on the same base parts, so it should be sufficient to define a dependency for the most specialised lib parts and let do ant/maven/gradle the actual work. 

however, the dependencies as listed in our configuration come from the libgdx creator itself. they should know what they are doing 🤔

cagix commented:
@AMatutat i'm not sure if we really need all the listed dependencies, e.g. do we really need libgdx-ai? could you check please and provide a hopefully shorter list to @Tobias321 ?

cagix commented:
@Tobias321 still merging master into features?

Tobias321 commented:
> @Tobias321 still merging master into features?

hi, sorry, that was the default option in GitHub Desktop... next time i will simply rebase the feature branch onto master.

now i have rebased the feature branch. if there is something terrible broken now, i could create a new "clean" feature branch `ant_ivy_port_2`. should i?

(and i need a beer. 😄 )

cagix commented:
.oO( use the source^Wterminal, luke )

seriously, i'd consider this a bad behaviour of github desktop. is there an option to change this configuration?

oh, here: 🍺 cheers :)

Tobias321 commented:
> is there an option to change this configuration?

you can choose:

![image](https://user-images.githubusercontent.com/85501570/122678967-2dbedb80-d1e9-11eb-8e5a-70b7d25e2c40.png)

cheers:)

Tobias321 commented:
please see #73

cagix commented:
> > is there an option to change this configuration?
> 
> you can choose:
> 
> ![image](https://user-images.githubusercontent.com/85501570/122678967-2dbedb80-d1e9-11eb-8e5a-70b7d25e2c40.png)
> 
> cheers:)

strange, github desktop recognises `master` as default branch, but still offers to merge it into feature branches. 

@Tobias321 you could have reset the last (merge) commit, rebase your branch and do a force push ...

Tobias321 commented:
> strange, github desktop recognises `master` as default branch, but still offers to merge it into feature branches.

if you clone a/your fork, github desktop ask you, what to work against:

![image](https://user-images.githubusercontent.com/85501570/122680163-3239c300-d1ee-11eb-93ea-ba929b4d3eda.png)

i have choose "For my own purposes", then your master is the default branch for a merge/rebase... 🙂

cagix commented:
> i have choose "For my own purposes", then your master is the default branch for a merge/rebase... 🙂

yeah, well, nevertheless. all merging should always go from feature into master, rebasing should always be feature on top on master ...

Tobias321 commented:
> > i have choose "For my own purposes", then your master is the default branch for a merge/rebase... 🙂
> 
> yeah, well, nevertheless. all merging should always go from feature into master, rebasing should always be feature on top on master ...

i'm quite not understand that completely.

is:
```
git checkout feature1
git rebase master
```
and
`git rebase --skip`
as mentioned here: https://stackoverflow.com/a/16956062
the best pattern/practice, to rebase a feature branch onto a hotfix branch (the own master here)?

cagix commented:
just like this. 

a more verbose explanation:
1. update your master with upstream's master:  
    * `git checkout master && git pull upstream master` (this should work without any problems provided you don't changed your master locally(!), but in such a scenario is changing the local master considered bad practice -- just use feature branches instead) => see https://git-scm.com/book/en/v2/Git-Branching-Remote-Branches
    * `git checkout master && git pull --rebase upstream master` (in case you changed your master locally -- note: you may want to push your new master to upstream, i.e. `push origin master` and create a pull request from `origin:master` to `upstream:master`)
2. rebase your own feature branch onto the master: `git rebase master feature` => see https://git-scm.com/book/en/v2/Git-Branching-Rebasing
3. if the feature branch has already been published (i.e. a pull request exists with it), then force push to update the pull request: `git push -f origin feature` (assuming that `origin` is the fork of `upstream` and your pull request is from `origin:feature` to `upstream:master`).

cagix commented:
so, you always need to synchronise your master with upstream's master, and then rebase your feature branch on top of the master. this way you always have the latest features and changes available for your feature branch and you avoid merge conflicts (mostly). 

push both branches (updates master and feature) to your fork and create a pull request from your fork's feature branch into upstream's master.

Tobias321 commented:
many thanks 🙂 👍🏻
